include/alpaka/rand/RandDefault.hpp

```diff
@@ -0,0 +1,226 @@
+
+                //! The GPU random number normal distribution.
+                template<typename T>
+                class UniformUint
+                {
+                    static_assert(std::is_integral<T>::value, "Return type of UniformUint must be integral.");
+
+                public:
+                    UniformUint() = default;
+
+                    template<typename TEngine>
+                    ALPAKA_FN_HOST_ACC auto operator()(TEngine& engine) -> T
+                    {
+                        using BitsT = typename TEngine::result_type;
+                        T ret = 0;
+                        constexpr auto N = sizeof(T) / sizeof(BitsT);
+                        for(unsigned int a = 0; a < N; ++a)
+                        {
+                            ret
+                                ^= (static_cast<T>(engine())
+                                    << (sizeof(BitsT) * std::numeric_limits<unsigned char>::digits * a));
+                        }
+                        return ret;
+                    }
+                };
+
+                //! The GPU random number uniform distribution.
+                template<typename T>
+                class UniformReal
+                {
+                    static_assert(
+                        std::is_floating_point<T>::value,
+                        "Return type of UniformReal must be floating point.");
+
+                    using BitsT = typename detail::BitsType<T>::type;
+
+                public:
+                    UniformReal() = default;
+
+                    template<typename TEngine>
+                    ALPAKA_FN_HOST_ACC auto operator()(TEngine& engine) -> T
+                    {
+                        constexpr BitsT limit = static_cast<BitsT>(1) << std::numeric_limits<T>::digits;
+                        const BitsT b = UniformUint<BitsT>()(engine);
+                        const auto ret = static_cast<T>(b & (limit - 1)) / limit;
+                        return ret;
+                    }
+                };
+
+                /*! The GPU random number normal distribution.
+                 *
+                 * \note
+                 * This type contains state and is not thread-safe: To be used
+                 * per thread, not shared.
+                 */
+                template<typename Acc, typename T>
+                class NormalReal
+                {
+                    static_assert(
+                        std::is_floating_point<T>::value,
+                        "Return type of NormalReal must be floating point.");
+
+                    const Acc& m_acc;
+                    T m_buf = std::numeric_limits<T>::quiet_NaN();
+
+                public:
+                    NormalReal(const Acc& acc) : m_acc(acc)
+                    {
+                    }
+
+                    //! The copy ctor does not copy `m_buf`.
+                    NormalReal(const NormalReal& o) : m_acc(o.m_acc)
+                    {
+                    }
```

Comment:
I think this behavior is surprising. If you take a copy of a `NormalReal` and then ask for the next number on the old and new instance, you will get different results.

I would prefer to have the default copy ctor here.

Reply:
This is only the distribution, the engine determines which numbers are generated. The expected behavior would be:
```c++
NormalReal<Acc> a(acc), b(acc);
Engine<Acc> e(acc);

inGeneral( a(e) != b(e) ) ; // because of two engine invocations

b = a;

inGeneral( a(e) != b(e) ) ; // because of two engine invocations
```

If a value was cached in `a` before `a = b` and the cache is not cleared, you break this behavior, because neither `a` nor `b` invoke the engine to ask for a new random number. This would not happen with a distribution that does not cache numbers. Caching is just an optimization, it must not interfere with the randomness.

Reply:
Ok, that's right! I could still make the case tough that you could copy the engine as well and expect the results to be the same, but maybe that is not a relevant use case. Anyway, I think it's fine then.

Reply:
You can either copy the cache or not, both have drawbacks, but the property I stated above is essential. If we give weight to your second case, the caching optimization not be possible. Copying the engine it a thing that might actually happen. One would expect a any distribution instance to give the same numbers on both engines, which also speaks against caching. However removing the cache and only using one of the two random numbers seems to sacrifice performance to support some very rare case. The way I see to minimize this problem is to disallow copy and move as far as possible (we only need the move ctor currently and can delete all when moving to C++17) and document the interaction of cache and engines. I pushed https://github.com/alpaka-group/alpaka/pull/1444/commits/004ae6cdd8f4a3321ecc0c0e22bf96cb9074204e along those lines.